File: src/vod_strm_builder/xtream.py
from __future__ import annotations

from typing import Any
from urllib.parse import quote

import requests

from .models import MovieItem, ProviderConfig, SeriesItem
from .utils import clean_title, extract_year


class XtreamClient:
    def __init__(self, provider: ProviderConfig, timeout: int = 45) -> None:
        self.provider = provider
        self.timeout = timeout
        self.session = requests.Session()
        self.session.headers.update({"User-Agent": provider.user_agent})

    def player_api(self, action: str) -> Any:
        url = f"{self.provider.server_url}/player_api.php"
        try:
            response = self.session.get(
                url,
                params={
                    "username": self.provider.username,
                    "password": self.provider.password,
                    "action": action,
                },
                timeout=self.timeout,
            )
            response.raise_for_status()
        except requests.RequestException:
            raise RuntimeError(f"Xtream player_api action '{action}' failed for configured server") from None
        return response.json()

    def categories(self, kind: str) -> dict[str, str]:
        action = "get_vod_categories" if kind == "movie" else "get_series_categories"
        rows = self.player_api(action)
        return {str(row.get("category_id")): str(row.get("category_name") or "") for row in rows or []}

    def movies(self) -> list[MovieItem]:
        rows = self.player_api("get_vod_streams") or []
        items: list[MovieItem] = []
        for row in rows:
            name = str(row.get("name") or row.get("title") or "").strip()
            stream_id = str(row.get("stream_id") or "")
            if not name or not stream_id:
                continue
            ext = str(row.get("container_extension") or "mp4").lstrip(".")
            tmdb_id = _clean_id(row.get("tmdb") or row.get("tmdb_id"))
            items.append(
                MovieItem(
                    name=name,
                    stream_id=stream_id,
                    category_id=str(row.get("category_id") or ""),
                    extension=ext,
                    year=extract_year(name, row.get("releaseDate"), row.get("release_date"), row.get("year")),
                    tmdb_id=tmdb_id,
                    imdb_id=_clean_id(row.get("imdb") or row.get("imdb_id")),
                    plot=_string_or_none(row.get("plot")),
                    genre=_string_or_none(row.get("genre")),
                    rating=_string_or_none(row.get("rating")),
                    cover=_string_or_none(row.get("stream_icon") or row.get("cover")),
                )
            )
        return items

    def series(self) -> list[SeriesItem]:
        rows = self.player_api("get_series") or []
        items: list[SeriesItem] = []
        for row in rows:
            name = str(row.get("name") or "").strip()
            series_id = str(row.get("series_id") or "")
            if not name or not series_id:
                continue
            tmdb_id = _clean_id(row.get("tmdb") or row.get("tmdb_id"))
            items.append(
                SeriesItem(
                    name=name,
                    series_id=series_id,
                    category_id=str(row.get("category_id") or ""),
                    year=extract_year(name, row.get("releaseDate"), row.get("release_date"), row.get("year")),
                    tmdb_id=tmdb_id,
                    imdb_id=_clean_id(row.get("imdb") or row.get("imdb_id")),
                    plot=_string_or_none(row.get("plot")),
                    genre=_string_or_none(row.get("genre")),
                    rating=_string_or_none(row.get("rating")),
                    cover=_string_or_none(row.get("cover")),
                )
            )
        return items

    def movie_url(self, item: MovieItem) -> str:
        return (
            f"{self.provider.server_url}/movie/"
            f"{quote(self.provider.username)}/{quote(self.provider.password)}/"
            f"{quote(item.stream_id)}.{item.extension or 'mp4'}"
        )

    def m3u_source(self) -> str:
        if self.provider.m3u_file:
            return str(self.provider.m3u_file)
        if self.provider.m3u_url:
            return self.provider.m3u_url
        return (
            f"{self.provider.server_url}/get.php?"
            f"username={quote(self.provider.username)}&password={quote(self.provider.password)}"
            "&type=m3u_plus&output=ts"
        )


def _clean_id(value: object) -> str | None:
    if value in (None, "", 0, "0"):
        return None
    text = str(value).strip()
    return text if text and text != "0" else None


def _string_or_none(value: object) -> str | None:
    if value in (None, ""):
        return None
    if isinstance(value, list):
        return ", ".join(str(item) for item in value if item)
    return str(value)

File: crates/vsb/src/xtream.rs
use anyhow::{Context, Result};
use reqwest::header::{HeaderMap, HeaderValue, USER_AGENT};
use serde_json::Value;

use crate::{
    config::{InputConfig, XtreamSeriesSource},
    m3u,
    model::{MediaItem, MediaKind},
    text::extract_year,
};

pub async fn load_input(input: &InputConfig, base_url: &str) -> Result<Vec<MediaItem>> {
    let client = XtreamClient::new(input, base_url)?;
    let mut items = Vec::new();
    if !input.options.xtream_skip_vod {
        items.extend(client.movies().await?);
    }
    if !input.options.xtream_skip_series {
        match input.options.xtream_series_source {
            XtreamSeriesSource::M3u => {
                let playlist_url = client.m3u_url();
                items.extend(
                    m3u::load_input(input, &playlist_url)
                        .await?
                        .into_iter()
                        .filter(|item| item.kind == MediaKind::Series),
                );
            }
            XtreamSeriesSource::Api => {
                items.extend(client.series().await?);
            }
        }
    }
    if !input.options.xtream_skip_live {
        items.extend(client.live().await?);
    }
    Ok(items)
}

struct XtreamClient<'a> {
    input: &'a InputConfig,
    client: reqwest::Client,
    base_url: String,
}

impl<'a> XtreamClient<'a> {
    fn new(input: &'a InputConfig, base_url: &str) -> Result<Self> {
        let mut headers = HeaderMap::new();
        headers.insert(
            USER_AGENT,
            HeaderValue::from_str(&input.options.user_agent)?,
        );
        for (key, value) in &input.headers {
            headers.insert(
                reqwest::header::HeaderName::from_bytes(key.as_bytes())?,
                HeaderValue::from_str(value)?,
            );
        }
        Ok(Self {
            input,
            client: reqwest::Client::builder()
                .default_headers(headers)
                .build()?,
            base_url: base_url.trim_end_matches('/').to_string(),
        })
    }

    async fn live(&self) -> Result<Vec<MediaItem>> {
        let categories = self.categories("get_live_categories").await?;
        let rows = self.player_api("get_live_streams", &[]).await?;
        Ok(rows
            .as_array()
            .into_iter()
            .flatten()
            .filter_map(|row| {
                let name = string_field(row, &["name", "title"])?;
                let stream_id = string_field(row, &["stream_id"])?;
                let category_id = string_field(row, &["category_id"]).unwrap_or_default();
                let group = categories.get(&category_id).cloned().unwrap_or(category_id);
                Some(MediaItem {
                    input: self.input.name.clone(),
                    kind: MediaKind::Live,
                    group,
                    name,
                    url: self.live_url(&stream_id),
                    stream_id: Some(stream_id),
                    extension: Some("ts".to_string()),
                    year: None,
                    tmdb_id: None,
                    logo: string_field(row, &["stream_icon", "cover"]),
                    episode: None,
                })
            })
            .collect())
    }

    async fn movies(&self) -> Result<Vec<MediaItem>> {
        let categories = self.categories("get_vod_categories").await?;
        let rows = self.player_api("get_vod_streams", &[]).await?;
        Ok(rows
            .as_array()
            .into_iter()
            .flatten()
            .filter_map(|row| {
                let name = string_field(row, &["name", "title"])?;
                let stream_id = string_field(row, &["stream_id"])?;
                let ext = string_field(row, &["container_extension"])
                    .unwrap_or_else(|| "mp4".to_string());
                let category_id = string_field(row, &["category_id"]).unwrap_or_default();
                let group = categories.get(&category_id).cloned().unwrap_or(category_id);
                Some(MediaItem {
                    input: self.input.name.clone(),
                    kind: MediaKind::Movie,
                    group,
                    year: extract_year(&name),
                    tmdb_id: clean_id(row.get("tmdb").or_else(|| row.get("tmdb_id"))),
                    logo: string_field(row, &["stream_icon", "cover"]),
                    url: self.movie_url(&stream_id, &ext),
                    stream_id: Some(stream_id),
                    extension: Some(ext),
                    name,
                    episode: None,
                })
            })
            .collect())
    }

    async fn series(&self) -> Result<Vec<MediaItem>> {
        let categories = self.categories("get_series_categories").await?;
        let rows = self.player_api("get_series", &[]).await?;
        let mut output = Vec::new();
        for row in rows.as_array().into_iter().flatten() {
            let Some(series_name) = string_field(row, &["name"]) else {
                continue;
            };
            let Some(series_id) = string_field(row, &["series_id"]) else {
                continue;
            };
            let category_id = string_field(row, &["category_id"]).unwrap_or_default();
            let group = categories.get(&category_id).cloned().unwrap_or(category_id);
            let info = self
                .player_api("get_series_info", &[("series_id", series_id.as_str())])
                .await?;
            for episode_row in iter_series_episode_rows(&info) {
                let Some(stream_id) = string_field(episode_row, &["id", "stream_id"]) else {
                    continue;
                };
                let season = numeric_field(episode_row, &["season"]).unwrap_or(1);
                let episode =
                    numeric_field(episode_row, &["episode_num", "episode", "num"]).unwrap_or(1);
                let ext = string_field(episode_row, &["container_extension"])
                    .unwrap_or_else(|| "mp4".to_string());
                let title = string_field(episode_row, &["title"])
                    .unwrap_or_else(|| format!("Episode {episode:02}"));
                output.push(MediaItem {
                    input: self.input.name.clone(),
                    kind: MediaKind::Series,
                    group: group.clone(),
                    name: series_name.clone(),
                    url: self.series_url(&stream_id, &ext),
                    stream_id: Some(stream_id),
                    extension: Some(ext),
                    year: extract_year(&series_name),
                    tmdb_id: clean_id(row.get("tmdb").or_else(|| row.get("tmdb_id"))),
                    logo: string_field(row, &["cover"]),
                    episode: Some(crate::model::EpisodeInfo {
                        series_name: series_name.clone(),
                        season,
                        episode,
                        title,
                    }),
                });
            }
        }
        Ok(output)
    }

    async fn categories(&self, action: &str) -> Result<std::collections::HashMap<String, String>> {
        let rows = self.player_api(action, &[]).await?;
        Ok(rows
            .as_array()
            .into_iter()
            .flatten()
            .filter_map(|row| {
                Some((
                    string_field(row, &["category_id"])?,
                    string_field(row, &["category_name"])?,
                ))
            })
            .collect())
    }

    async fn player_api(&self, action: &str, extra: &[(&str, &str)]) -> Result<Value> {
        let username = self.input.username.as_deref().unwrap_or_default();
        let password = self.input.password.as_deref().unwrap_or_default();
        let mut query = vec![
            ("username", username),
            ("password", password),
            ("action", action),
        ];
        query.extend_from_slice(extra);
        let response = self
            .client
            .get(format!("{}/player_api.php", self.base_url))
            .query(&query)
            .send()
            .await
            .with_context(|| format!("xtream action {action} for {}", self.input.name))?;
        let status = response.status();
        if !status.is_success() {
            anyhow::bail!("xtream action {action} failed with HTTP {status}");
        }
        response
            .json::<Value>()
            .await
            .with_context(|| format!("parse xtream action {action} JSON"))
    }

    fn m3u_url(&self) -> String {
        format!(
            "{}/get.php?username={}&password={}&type=m3u_plus&output=ts",
            self.base_url,
            urlencoding::encode(self.input.username.as_deref().unwrap_or_default()),
            urlencoding::encode(self.input.password.as_deref().unwrap_or_default())
        )
    }

    fn live_url(&self, stream_id: &str) -> String {
        format!(
            "{}/live/{}/{}/{}.ts",
            self.base_url,
            urlencoding::encode(self.input.username.as_deref().unwrap_or_default()),
            urlencoding::encode(self.input.password.as_deref().unwrap_or_default()),
            urlencoding::encode(stream_id)
        )
    }

    fn movie_url(&self, stream_id: &str, ext: &str) -> String {
        format!(
            "{}/movie/{}/{}/{}.{}",
            self.base_url,
            urlencoding::encode(self.input.username.as_deref().unwrap_or_default()),
            urlencoding::encode(self.input.password.as_deref().unwrap_or_default()),
            urlencoding::encode(stream_id),
            ext.trim_start_matches('.')
        )
    }

    fn series_url(&self, stream_id: &str, ext: &str) -> String {
        format!(
            "{}/series/{}/{}/{}.{}",
            self.base_url,
            urlencoding::encode(self.input.username.as_deref().unwrap_or_default()),
            urlencoding::encode(self.input.password.as_deref().unwrap_or_default()),
            urlencoding::encode(stream_id),
            ext.trim_start_matches('.')
        )
    }
}

fn string_field(row: &Value, keys: &[&str]) -> Option<String> {
    keys.iter()
        .find_map(|key| row.get(key))
        .and_then(|value| match value {
            Value::String(text) => Some(text.trim().to_string()),
            Value::Number(number) => Some(number.to_string()),
            _ => None,
        })
        .filter(|value| !value.is_empty() && value != "0")
}

fn numeric_field(row: &Value, keys: &[&str]) -> Option<u16> {
    keys.iter()
        .find_map(|key| row.get(key))
        .and_then(|value| {
            value
                .as_u64()
                .or_else(|| value.as_str()?.parse::<u64>().ok())
        })
        .and_then(|value| u16::try_from(value).ok())
}

fn clean_id(value: Option<&Value>) -> Option<String> {
    value
        .and_then(|value| match value {
            Value::String(text) => Some(text.trim().to_string()),
            Value::Number(number) => Some(number.to_string()),
            _ => None,
        })
        .filter(|value| !value.is_empty() && value != "0")
}

fn iter_series_episode_rows(info: &Value) -> Vec<&Value> {
    let Some(episodes) = info.get("episodes").and_then(Value::as_object) else {
        return Vec::new();
    };
    episodes
        .values()
        .flat_map(|season| season.as_array().into_iter().flatten())
        .collect()
}
